### PR TITLE
Make ser/de configurable in checkpointer classes

### PR DESCRIPTION
**Pull request by nfcampos**
- Remove pydantic usage from base checkpointer class
- Make serialization configurable for all existing checkpointer classes, you can eg use dill or json instead of pickle

### CODE REVIEW

File: langgraph/checkpoint/memory.py [hunk 1]
```diff
@@ -1,47 +1,84 @@
...
-                            "thread_id": config["configurable"]["thread_id"],
-                            "thread_ts": thread_ts,
-                        }
-                    },
-                    checkpoint=checkpoints[thread_ts],
+                    config={"configurable": {"thread_id": thread_id, "thread_ts": ts}},
+                    checkpoint=self.serde.loads(checkpoints[ts]),
                 )
 
+    def list(self, config: RunnableConfig) -> Iterator[CheckpointTuple]:
```
Comment (nfcampos) [inline langgraph/checkpoint/memory.py:46]: this was missing

File: langgraph/checkpoint/base.py [hunk 2]
```diff
@@ -93,8 +98,27 @@ class CheckpointTuple(NamedTuple):
...
+        ...
+
+    def loads(self, data: bytes) -> Any:
+        ...
+
+
+class BaseCheckpointSaver(ABC):
+    at: CheckpointAt = CheckpointAt.END_OF_STEP
+
+    serde: SerializerProtocol
```
Comment (hinthornw) [inline langgraph/checkpoint/base.py:112]: so this is for specifying an inheritable default?

File: langgraph/checkpoint/memory.py [hunk 3]
```diff
@@ -1,47 +1,84 @@
...
-        if value := self.get_tuple(config):
-            return value.checkpoint
+    storage: defaultdict[str, dict[str, Checkpoint]]
+
+    def __init__(
+        self,
+        *,
+        serde: Optional[SerializerProtocol] = None,
+        at: Optional[CheckpointAt] = None,
+    ) -> None:
```
Comment (hinthornw) [inline langgraph/checkpoint/memory.py:27]: Could we add docstrings :)

**Review by hinthornw — APPROVED** [review]

File: langgraph/checkpoint/memory.py (hunk 3 above)
Comment (nfcampos) [reply langgraph/checkpoint/memory.py:27]: haha yea but only after settling on final checkpointer interface

File: langgraph/checkpoint/base.py [hunk 4]
```diff
@@ -93,8 +98,27 @@ class CheckpointTuple(NamedTuple):
...
+class SerializerProtocol(Protocol):
+    def dumps(self, obj: Any) -> bytes:
+        ...
+
+    def loads(self, data: bytes) -> Any:
+        ...
+
+
+class BaseCheckpointSaver(ABC):
+    at: CheckpointAt = CheckpointAt.END_OF_STEP
```
Comment (andrewnguonly) [inline langgraph/checkpoint/base.py:110]: More granular default check pointing.

File: langgraph/checkpoint/base.py (hunk 2 above)
Comment (nfcampos) [reply langgraph/checkpoint/base.py:112]: you mean the previous line?

**Review by andrewnguonly — APPROVED** [review]
High level, makes sense!

**Pull request merged**